From b2013abeb80ff17b042ead6c6164c975f7778dd3 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?R=C3=A9mi=20Bernon?= <rbernon@codeweavers.com>
Date: Wed, 31 Mar 2021 10:33:04 +0200
Subject: [PATCH 1/2] winex11.drv: Use bottom-most desktop window type for
 dummy parent.

Instead of override-redirect, whenever the window manager reports
support for the atoms.
---
 dlls/winex11.drv/window.c      | 29 +++++++++++++++++++++++++++++
 dlls/winex11.drv/x11drv.h      |  6 ++++++
 dlls/winex11.drv/x11drv_main.c | 10 ++++++++++
 3 files changed, 45 insertions(+)

diff --git a/dlls/winex11.drv/window.c b/dlls/winex11.drv/window.c
index 11111111111..11111111111 100644
--- a/dlls/winex11.drv/window.c
+++ b/dlls/winex11.drv/window.c
@@ -1544,17 +1544,46 @@ static void move_window_bits( HWND hwnd, Window window, const RECT *old_rect, co
 Window get_dummy_parent(void)
 {
     static Window dummy_parent;
+    XWMHints *hints = NULL;
+    DWORD count;
+    Atom atoms[3];
 
     if (!dummy_parent)
     {
         XSetWindowAttributes attrib;
 
         attrib.override_redirect = True;
+        if (ewmh.has__net_wm_state && ewmh.has__net_wm_state_below &&
+            ewmh.has__net_wm_window_type && ewmh.has__net_wm_window_type_desktop &&
+            (hints = XAllocWMHints()))
+            attrib.override_redirect = False;
+
         attrib.border_pixel = 0;
         attrib.colormap = default_colormap;
         dummy_parent = XCreateWindow( gdi_display, root_window, -1, -1, 1, 1, 0, default_visual.depth,
                                       InputOutput, default_visual.visual,
                                       CWColormap | CWBorderPixel | CWOverrideRedirect, &attrib );
+
+        if (!attrib.override_redirect)
+        {
+            count = 0;
+            atoms[count++] = x11drv_atom(_NET_WM_STATE_BELOW);
+            atoms[count++] = x11drv_atom(_NET_WM_STATE_SKIP_TASKBAR);
+            atoms[count++] = x11drv_atom(_NET_WM_STATE_SKIP_PAGER);
+            XChangeProperty( gdi_display, dummy_parent, x11drv_atom(_NET_WM_STATE), XA_ATOM,
+                             32, PropModeReplace, (unsigned char *)atoms, count );
+
+            count = 0;
+            atoms[count++] = x11drv_atom(_NET_WM_WINDOW_TYPE_DESKTOP);
+            XChangeProperty( gdi_display, dummy_parent, x11drv_atom(_NET_WM_WINDOW_TYPE), XA_ATOM,
+                             32, PropModeReplace, (unsigned char *)atoms, count );
+
+            hints->flags = InputHint;
+            hints->input = False;
+            XSetWMHints( gdi_display, dummy_parent, hints );
+            XFree( hints );
+        }
+
         XMapWindow( gdi_display, dummy_parent );
     }
     return dummy_parent;
diff --git a/dlls/winex11.drv/x11drv.h b/dlls/winex11.drv/x11drv.h
index 11111111111..11111111111 100644
--- a/dlls/winex11.drv/x11drv.h
+++ b/dlls/winex11.drv/x11drv.h
@@ -418,6 +418,10 @@ struct x11drv_ewmh_data
 {
     int has__net_active_window : 1;
     int has__net_wm_moveresize : 1;
+    int has__net_wm_state : 1;
+    int has__net_wm_state_below : 1;
+    int has__net_wm_window_type : 1;
+    int has__net_wm_window_type_desktop : 1;
 };
 
 extern XVisualInfo default_visual DECLSPEC_HIDDEN;
@@ -498,6 +502,7 @@ enum x11drv_atoms
     XATOM__NET_WM_PING,
     XATOM__NET_WM_STATE,
     XATOM__NET_WM_STATE_ABOVE,
+    XATOM__NET_WM_STATE_BELOW,
     XATOM__NET_WM_STATE_DEMANDS_ATTENTION,
     XATOM__NET_WM_STATE_FULLSCREEN,
     XATOM__NET_WM_STATE_MAXIMIZED_HORZ,
@@ -508,6 +513,7 @@ enum x11drv_atoms
     XATOM__NET_WM_USER_TIME_WINDOW,
     XATOM__NET_WM_WINDOW_OPACITY,
     XATOM__NET_WM_WINDOW_TYPE,
+    XATOM__NET_WM_WINDOW_TYPE_DESKTOP,
     XATOM__NET_WM_WINDOW_TYPE_DIALOG,
     XATOM__NET_WM_WINDOW_TYPE_NORMAL,
     XATOM__NET_WM_WINDOW_TYPE_UTILITY,
diff --git a/dlls/winex11.drv/x11drv_main.c b/dlls/winex11.drv/x11drv_main.c
index 11111111111..11111111111 100644
--- a/dlls/winex11.drv/x11drv_main.c
+++ b/dlls/winex11.drv/x11drv_main.c
@@ -177,6 +177,7 @@ static const char * const atom_names[NB_XATOMS - FIRST_XATOM] =
     "_NET_WM_PING",
     "_NET_WM_STATE",
     "_NET_WM_STATE_ABOVE",
+    "_NET_WM_STATE_BELOW",
     "_NET_WM_STATE_DEMANDS_ATTENTION",
     "_NET_WM_STATE_FULLSCREEN",
     "_NET_WM_STATE_MAXIMIZED_HORZ",
@@ -187,6 +188,7 @@ static const char * const atom_names[NB_XATOMS - FIRST_XATOM] =
     "_NET_WM_USER_TIME_WINDOW",
     "_NET_WM_WINDOW_OPACITY",
     "_NET_WM_WINDOW_TYPE",
+    "_NET_WM_WINDOW_TYPE_DESKTOP",
     "_NET_WM_WINDOW_TYPE_DIALOG",
     "_NET_WM_WINDOW_TYPE_NORMAL",
     "_NET_WM_WINDOW_TYPE_UTILITY",
@@ -571,6 +573,14 @@ static void x11drv_ewmh_init(void)
             ewmh.has__net_active_window = 1;
         if (supported[i] == x11drv_atom(_NET_WM_MOVERESIZE))
             ewmh.has__net_wm_moveresize = 1;
+        if (supported[i] == x11drv_atom(_NET_WM_STATE))
+            ewmh.has__net_wm_state = 1;
+        if (supported[i] == x11drv_atom(_NET_WM_STATE_BELOW))
+            ewmh.has__net_wm_state_below = 1;
+        if (supported[i] == x11drv_atom(_NET_WM_WINDOW_TYPE))
+            ewmh.has__net_wm_window_type = 1;
+        if (supported[i] == x11drv_atom(_NET_WM_WINDOW_TYPE_DESKTOP))
+            ewmh.has__net_wm_window_type_desktop = 1;
 
         atom_name = XGetAtomName( gdi_display, supported[i] );
         TRACE( "  %s\n", atom_name );
-- 
2.37.2

